### PR TITLE
Bring user straight to reader comments s creen when notification comment is tapped

### LINKED ISSUE #22515: 🍎 Bring the user straight to the comments reader view
Non-admin users that cannot moderate comment should skip the comment details screen and be directed straight to the comments reader view

### PR DESCRIPTION
**Pull request by salimbraksa**
Fixes #22515

## Description

This PR makes the following changes when a notification is tapped:
- Non-admin users are navigated directly to reader comments screen.
- For admin users the old comment detail screen is shown.

## Test Instructions

**Simulate non-admin user**
1. Open `Comment+CoreDataClass.swift` file
2. Locate `allowsModeration` method.
3. Force it to return `false`

### Non-admin user
1. Simulate a non-admin user.
2. Build and run Jeptack app.
3. Navigate to Notifications.
4. Tap "Comments" tab.
5. Tap a comment notification.
6. **Expect** the Reader Comments screen to appear with the comment highlighted.
7. Tap "Up" or "Down" arrows.
8. Expect the next comment to load.

### Admin user ( Regression )
1. Navigate to Notifications.
2. Tap "Comments" tab.
3. Tap a comment notification.
4. **Expect** the Comment Detail screen to appear.
5. Tap "Up" or "Down" arrows.
6. Expect the next comment to load.

## Regression Notes
1. Potential unintended areas of impact
Users who have permission to moderate the comment should see the Comment Detail screen.

2. What I did to test those areas of impact (or what existing automated tests I relied on)
Manual testing.

3. What automated tests I added (or what prevented me from doing so)
None.

## PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding unit tests for my changes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

## Testing checklist:
- [ ] WordPress.com sites and self-hosted Jetpack sites.
- [ ] Portrait and landscape orientations.
- [ ] Light and dark modes.
- [ ] Fonts: Larger, smaller and bold text.
- [ ] High contrast.
- [ ] VoiceOver.
- [ ] Languages with large words or with letters/accents not frequently used in English.
- [ ] Right-to-left languages. (Even if translation isn’t complete, formatting should still respect the right-to-left layout)
- [ ] iPhone and iPad. 
- [ ] Multi-tasking: Split view and Slide over. (iPad)

### CODE REVIEW

**Comment by dangermattic** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          2 Warnings
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">View files have been modified, but no screenshot is included in the pull request. Consider adding some for clarity.</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">PR is not assigned to a milestone.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

File: WordPress/Classes/ViewRelated/Notifications/Controllers/NotificationCommentDetailViewController.swift [hunk 1]
```diff
@@ -156,21 +164,41 @@ private extension NotificationCommentDetailViewController {
...
+            let newContent = makeNewContent(with: comment)
+            let viewController = newContent.viewController
+            viewController.view.translatesAutoresizingMaskIntoConstraints = false
+            self.add(viewController)
+            self.view.pinSubviewToAllEdges(viewController.view)
+            self.content = newContent
         }
 
-        configureNavBarButtons()
+        // Configure navigation bar buttons
```
Comment (alpavanoglu) [inline WordPress/Classes/ViewRelated/Notifications/Controllers/NotificationCommentDetailViewController.swift]: Nit: This comment looks redundant as the function name is quite clear.

File: WordPress/Classes/ViewRelated/Notifications/Controllers/NotificationCommentDetailViewController.swift [hunk 2]
```diff
@@ -31,7 +33,13 @@ class NotificationCommentDetailViewController: UIViewController, NoResultsViewHo
     // In this case, use the Post to obtain Comment information.
     private var post: ReaderPost?
 
-    private var commentDetailViewController: CommentDetailViewController?
+    private var commentDetailViewController: CommentDetailViewController? {
+        guard let content, case let .commentDetails(viewController) = content  else {
+            return nil
+        }
+        return viewController
+    }
```
Comment (alpavanoglu) [inline WordPress/Classes/ViewRelated/Notifications/Controllers/NotificationCommentDetailViewController.swift:41]: It looks like `commentDetailViewController` is accessed only with an else statement. I suppose we simply switch on those cases and not need this value? Unless it is used commonly outside of changed lines

File: WordPress/Classes/ViewRelated/Notifications/Controllers/NotificationCommentDetailViewController.swift (hunk 1 above)
Comment (salimbraksa) [reply WordPress/Classes/ViewRelated/Notifications/Controllers/NotificationCommentDetailViewController.swift]: Resolved in 67958ebd6eeb96f9c7deb5771baed43aaf0bc2aa

File: WordPress/Classes/ViewRelated/Reader/Comments/ReaderCommentsViewController.swift [hunk 3]
```diff
@@ -32,24 +32,14 @@ extension NSNotification.Name {
...
-        cell.backgroundColor = .systemBackground
-        cell.configure(for: .thread, subtitle: post.titleForDisplay(), showsDisclosureIndicator: allowsPushingPostDetails)
-        cell.addGestureRecognizer(UITapGestureRecognizer(target: self, action: #selector(handleHeaderTapped)))
-
-        // the table view does not render separators for the section header views, so we need to create one.
-        cell.contentView.addBottomBorder(withColor: .separator, leadingMargin: tableView.separatorInset.left)
-
-        return cell
+        return headerView
     }
```
Comment (salimbraksa) [inline WordPress/Classes/ViewRelated/Reader/Comments/ReaderCommentsViewController.swift:43]: The reason behind this change is that sometimes the existing table header view disappears ( see screenshots below ). I was going to fix this issue but I found there is a new header view hidden behind a feature flag. I've tested it, and it works well, but not sure why the feature flag never got enabled. cc @dvdchr 

| Expected | Actual | New Header |
| --------- | ------- |  ----------- |
|  |  |   |

File: WordPress/Classes/ViewRelated/Reader/Comments/ReaderCommentsViewController.swift [hunk 4]
```diff
@@ -32,24 +32,14 @@ extension NSNotification.Name {
...
-                                                    showsDisclosureIndicator: true)
-            headerView.addGestureRecognizer(UITapGestureRecognizer(target: self, action: #selector(handleHeaderTapped)))
-            return headerView
+        let headerView = CommentTableHeaderView(
+            title: post.titleForDisplay(),
+            subtitle: .commentThread,
+            showsDisclosureIndicator: allowsPushingPostDetails
+        ) { [weak self] in
+            self?.handleHeaderTapped()
         }
```
Comment (salimbraksa) [inline WordPress/Classes/ViewRelated/Reader/Comments/ReaderCommentsViewController.swift:41]: @dvdchr I replaced the tap gesture with a closure-based action. [I also wrapped the whole SwiftUI](https://github.com/wordpress-mobile/WordPress-iOS/pull/22595/files#diff-2d42f23c24ce2e1b94bac9868b1341c1ffe124add3603221a7e97798d0c84890R105) view in a `Button`. The main difference between the Button and Tap Gesture, is that the Button _animates_ when tapped.

File: WordPress/Classes/ViewRelated/Comments/CommentDetailViewController.swift [hunk 5]
```diff
@@ -390,7 +390,7 @@ private extension CommentDetailViewController {
         return rows
     }
 
-    func configureModeratationRows() -> [RowType] {
+    func configureModerationRows() -> [RowType] {
```
Comment (dvdchr) [inline WordPress/Classes/ViewRelated/Comments/CommentDetailViewController.swift:393]: ❤️

File: WordPress/Classes/ViewRelated/Reader/Comments/ReaderCommentsViewController.swift (hunk 4 above)
Comment (dvdchr) [reply WordPress/Classes/ViewRelated/Reader/Comments/ReaderCommentsViewController.swift:41]: LGTM. 👍🏼

File: WordPress/Classes/ViewRelated/Reader/Comments/ReaderCommentsViewController.swift (hunk 3 above)
Comment (salimbraksa) [reply WordPress/Classes/ViewRelated/Reader/Comments/ReaderCommentsViewController.swift:43]: @dvdchr Thanks for clarifying! I can help out by opening a PR to update the comment header on the remaining screens.

Comment (dvdchr) [reply WordPress/Classes/ViewRelated/Reader/Comments/ReaderCommentsViewController.swift:43]: Oh, snap. I've mistakenly deleted my comment! I thought I double-posted... 😅  In essence: 

> The new comment header was part of #20568, which was an unplanned/unscheduled project that I took on during my downtime. The new design was supposed to be implemented on all three surfaces (site comments, comment threads, and comment notifications), but I only managed to ship for the comment threads. The header feature flag was disabled for consistency. Next week is our team's Maintenance Week, so I should be able to allocate some time to implement this for Site Comments.

File: WordPress/Classes/ViewRelated/Notifications/Controllers/NotificationCommentDetailViewController.swift (hunk 2 above)
Comment (salimbraksa) [reply WordPress/Classes/ViewRelated/Notifications/Controllers/NotificationCommentDetailViewController.swift:41]: Resolved internally

**Comment by alpavanoglu** [comment]
## 🧪  Test Results 🧪 

### Non-admin user
✅ Expect the Reader Comments screen to appear with the comment highlighted.
❓ The comment isn't positioned very well in my opinion. It looks centered however when the comment is relatively long, most of it is out of the screen. Wouldn't it better if the comment was aligned to top?

### Admin user ( Regression )
✅ Expect the next comment to load.

**Review by alpavanoglu — APPROVED** [review]
I left a possible improvement comment but it is non-blocking.

**Comment by salimbraksa** [comment]
>  The comment isn't positioned very well in my opinion. It looks centered however when the comment is relatively long, most of it is out of the screen. Wouldn't it better if the comment was aligned to top?

@alpavanoglu Is this issue particular to this PR or reproducible when navigating to the Reader Comments screen from the Comment Detail screen?

1. Login as Admin User
2. Notifications > Comment Notification 
3. Tap Header
4. The Reader Comments screen should appear

File: WordPress/Classes/ViewRelated/Reader/Comments/ReaderCommentsViewController.swift (hunk 3 above)
Comment (hassaanelgarem) [reply WordPress/Classes/ViewRelated/Reader/Comments/ReaderCommentsViewController.swift:43]: I see we've removed the feature flag check. Does this mean we no longer need it?

File: WordPress/Classes/ViewRelated/Notifications/Controllers/NotificationCommentDetailViewController.swift [hunk 6]
```diff
@@ -156,21 +164,41 @@ private extension NotificationCommentDetailViewController {
...
+            self.view.pinSubviewToAllEdges(viewController.view)
+            self.content = newContent
         }
 
-        configureNavBarButtons()
+        self.configureNavBarButtons()
+    }
+
+    private func makeNewContent(with comment: Comment) -> Content {
+        if let blog = comment.blog,
```
Comment (hassaanelgarem) [inline WordPress/Classes/ViewRelated/Notifications/Controllers/NotificationCommentDetailViewController.swift]: This If statement seems a bit complex and hard to follow. Is there a way we can simplify it?

File: WordPress/Classes/ViewRelated/Notifications/Controllers/NotificationCommentDetailViewController.swift [hunk 7]
```diff
@@ -326,4 +354,19 @@ private extension NotificationCommentDetailViewController {
         static let imageName = "wp-illustration-notifications"
     }
 
+    // MARK: - Types
+
+    private enum Content {
```
Comment (hassaanelgarem) [inline WordPress/Classes/ViewRelated/Notifications/Controllers/NotificationCommentDetailViewController.swift:375]: [Nit] Some documentation here would be great 🙏

**Review by hassaanelgarem — COMMENTED** [review]
I left some minor comments on the code, but they're not blockers. However, I noticed the following issues:
* Moving through comments using the up and down arrows doesn't work as expected
    1. Run the app
    2. Navigate to Notifications
    3. Open a comment that allow moderation
    4. Navigate to a comment that doesn't allow moderation through the up and down arrows
    5. Notice that the old view is shown while the expectation is to display the comment within the rest of the comments directly.
* Reply comments on sites where I'm not an admin or contributor do not work as expected
    1. Comment on a site that you're not a member of (Neither admin nor contributor)
    2. Reply to that comment from a different account
    3. Run the app
    4. Navigate to Notifications
    5. Open the notification from that last comment in step 2
    6. Notice that the old view is shown while the expectation is to display the comment within the rest of the comments directly.

**Comment by salimbraksa** [comment]
> Moving through comments using the up and down arrows doesn't work as expected

I was aware that this issue could arise if moderation was specific to each comment. However, I didn't address it initially because I assumed that wasn't the case—I believed that if a user had the ability to moderate one comment, they would have the same ability for others.

Now, it's clear to me: a comment might come from a site where the user has admin rights, while another could be from a site where they do not. Therefore, moderation indeed varies from comment to comment.

File: WordPress/Classes/ViewRelated/Reader/Comments/ReaderCommentsViewController.swift (hunk 3 above)
Comment (salimbraksa) [reply WordPress/Classes/ViewRelated/Reader/Comments/ReaderCommentsViewController.swift:43]: Personally I don't think we need it, I'm considering this header redesign similar to the redesigns we are doing now, which are not hidden behind a feature flag. Is that okey with you @dvdchr ?

File: WordPress/Classes/ViewRelated/Notifications/Controllers/NotificationCommentDetailViewController.swift (hunk 6 above)
Comment (salimbraksa) [reply WordPress/Classes/ViewRelated/Notifications/Controllers/NotificationCommentDetailViewController.swift]: I feel the same, I'll see how to improve readability. 🧘

Comment (salimbraksa) [reply WordPress/Classes/ViewRelated/Notifications/Controllers/NotificationCommentDetailViewController.swift]: Resolved in 7b0677405d7e8daad522222ee36ca6e18cdeb44b

**Comment by salimbraksa** [comment]
> Moving through comments using the up and down arrows doesn't work as expected

Resolved in 028df2413dc6c41a0c46067f37b45bfa900fe523

File: WordPress/Classes/ViewRelated/Notifications/Controllers/NotificationCommentDetailViewController.swift (hunk 7 above)
Comment (salimbraksa) [reply WordPress/Classes/ViewRelated/Notifications/Controllers/NotificationCommentDetailViewController.swift:375]: Resolved in 1e2d7b7362b4854571094d2106d6c3bfc3c87564

**Comment by salimbraksa** [comment]
> Reply comments on sites where I'm not an admin or contributor do not work as expected

Resolved in f56215d03193b2216b43c728ef8832c50c941055

File: WordPress/Classes/ViewRelated/Notifications/Controllers/NotificationCommentDetailViewController.swift [hunk 8]
```diff
@@ -166,11 +166,11 @@ private extension NotificationCommentDetailViewController {
 
         // Refresh the current content if the underlying view controller supports it
         // Else, remove the existing child view controller and add a new one.
-        if let commentDetailViewController {
+        if let commentDetailViewController, comment.allowsModeration() {
```
Comment (hassaanelgarem) [inline WordPress/Classes/ViewRelated/Notifications/Controllers/NotificationCommentDetailViewController.swift]: Instead of just relying on `comment.allowsModeration()` to decide if we should reload the view or not, using the content feels more deterministic especially if we add more conditions in the future.
```suggestion
        let newContent = makeNewContent(with: comment, notification: notification)
        if let commentDetailViewController, case .commentDetails = newContent {
            commentDetailViewController.refreshView(comment: comment, notification: notification)
        } else {
            self.content?.viewController.remove()
            let viewController = newContent.viewController
            ...
        }
```

**Review by hassaanelgarem — APPROVED** [review]
I tested the changes and they work as expected 🚀 
I just left one comment on the code 👍

**Review by alpavanoglu — APPROVED** [review]
The alignment issue seem to happen on `trunk` as well. So nohting this PR introduced.

File: WordPress/Classes/ViewRelated/Reader/Comments/ReaderCommentsViewController.swift (hunk 3 above)
Comment (dvdchr) [reply WordPress/Classes/ViewRelated/Reader/Comments/ReaderCommentsViewController.swift:43]: Sorry for the late reply, but yes feel free to remove it. The feature flag was intended for the larger comments moderation update work, but we never really got the green light to work on it. Might as well remove it and will properly restart it when we can "officially" work on it again! 🙂
